File: web/src/lib/utilities/capabilities.ts
/*
 * The decode capability probe (D-058).
 *
 * The plan request is made before the client knows the file's streams, so
 * the client cannot answer per file. It answers per *vocabulary* instead:
 * it probes a fixed matrix once and reports the tokens it could verify,
 * which the server validates against its own whitelist and applies to the
 * probed streams. A token is either a container ("mkv": the container
 * opens) or a container/codec pair ("mkv/h264": that codec family decodes
 * inside that container) — the pairing is what stops "HEVC decodes in
 * MP4" from licensing "HEVC in Matroska", which is a black screen.
 *
 * Two a-priori queries disagree, each in a different direction (measured
 * on Chrome 152): `decodingInfo` is right about HEVC and AC-3 but returns
 * a false negative for VP9+Opus, which in fact plays; `canPlayType` with
 * a codec string cannot see a platform decoder it has no string for. So a
 * token is reported when *either* says yes. The cost of a false positive
 * is one verification round-trip — the player proves the first decoded
 * frame and falls back to a transcode — while the cost of a false
 * negative is re-encoding a file the browser could have played.
 *
 * The matrix must stay in step with internal/contracts/capabilities.go:
 * a token the server does not know is dropped there, so drift costs
 * direct play silently. The browser smoke test is the guard.
 */

/** One codec the client tries to prove, with its RFC 6381 string. */
interface CodecProbe {
	token: string;
	/** null when the container's codec takes no codecs parameter. */
	codec: string | null;
}

interface ContainerProbe {
	token: string;
	/** Empty when the container carries no video (mp3, flac). */
	videoMime: string;
	audioMime: string;
	videoCodecs: CodecProbe[];
	audioCodecs: CodecProbe[];
}

const VIDEO_CODECS: Record<string, CodecProbe> = {
	h264: { token: 'h264', codec: 'avc1.64001f' },
	hevc: { token: 'hevc', codec: 'hvc1.1.6.L93.B0' },
	vp8: { token: 'vp8', codec: 'vp8' },
	vp9: { token: 'vp9', codec: 'vp09.00.10.08' },
	av1: { token: 'av1', codec: 'av01.0.04M.08' }
};

const AUDIO_CODECS: Record<string, CodecProbe> = {
	aac: { token: 'aac', codec: 'mp4a.40.2' },
	mp3: { token: 'mp3', codec: 'mp3' },
	opus: { token: 'opus', codec: 'opus' },
	vorbis: { token: 'vorbis', codec: 'vorbis' },
	flac: { token: 'flac', codec: 'flac' },
	ac3: { token: 'ac3', codec: 'ac-3' },
	eac3: { token: 'eac3', codec: 'ec-3' }
};

const MATRIX: ContainerProbe[] = [
	{
		token: 'mp4',
		videoMime: 'video/mp4',
		audioMime: 'audio/mp4',
		videoCodecs: [VIDEO_CODECS.h264, VIDEO_CODECS.hevc, VIDEO_CODECS.av1, VIDEO_CODECS.vp9],
		audioCodecs: [
			AUDIO_CODECS.aac,
			AUDIO_CODECS.mp3,
			AUDIO_CODECS.ac3,
			AUDIO_CODECS.eac3,
			AUDIO_CODECS.opus,
			AUDIO_CODECS.flac
		]
	},
	{
		token: 'mkv',
		videoMime: 'video/x-matroska',
		audioMime: 'audio/x-matroska',
		videoCodecs: [VIDEO_CODECS.h264, VIDEO_CODECS.hevc, VIDEO_CODECS.vp9, VIDEO_CODECS.av1],
		audioCodecs: [
			AUDIO_CODECS.aac,
			AUDIO_CODECS.mp3,
			AUDIO_CODECS.opus,
			AUDIO_CODECS.vorbis,
			AUDIO_CODECS.flac,
			AUDIO_CODECS.ac3,
			AUDIO_CODECS.eac3
		]
	},
	{
		token: 'webm',
		videoMime: 'video/webm',
		audioMime: 'audio/webm',
		videoCodecs: [VIDEO_CODECS.vp8, VIDEO_CODECS.vp9, VIDEO_CODECS.av1],
		audioCodecs: [AUDIO_CODECS.opus, AUDIO_CODECS.vorbis]
	},
	{
		token: 'ogg',
		videoMime: 'video/ogg',
		audioMime: 'audio/ogg',
		videoCodecs: [],
		audioCodecs: [AUDIO_CODECS.opus, AUDIO_CODECS.vorbis, AUDIO_CODECS.flac]
	},
	{
		token: 'mp3',
		videoMime: '',
		audioMime: 'audio/mpeg',
		videoCodecs: [],
		audioCodecs: [{ token: 'mp3', codec: null }]
	},
	{
		token: 'flac',
		videoMime: '',
		audioMime: 'audio/flac',
		videoCodecs: [],
		audioCodecs: [{ token: 'flac', codec: null }]
	}
];

/** The two questions a browser can answer about one configuration. */
export interface CapabilityOracles {
	/** `canPlayType`, possibly with a codecs parameter. */
	canPlay: (type: string) => string;
	/** `navigator.mediaCapabilities.decodingInfo().supported`. */
	decodingSupported: (mime: string, codec: string | null) => Promise<boolean>;
}

/*
 * The pure half of the probe: given the two oracles a browser offers,
 * which tokens does this client report? Exported so the matrix and the
 * union rule are testable without a browser.
 */
export async function tokensFor(oracles: CapabilityOracles): Promise<string[]> {
	const tokens: string[] = [];
	for (const container of MATRIX) {
		const opens = [container.videoMime, container.audioMime].some(
			(mime) => mime !== '' && oracles.canPlay(mime) !== ''
		);
		if (!opens) continue;
		tokens.push(container.token);
		for (const side of [
			{ mime: container.videoMime, codecs: container.videoCodecs },
			{ mime: container.audioMime, codecs: container.audioCodecs }
		]) {
			if (side.mime === '') continue;
			for (const codec of side.codecs) {
				const type = codec.codec === null ? side.mime : `${side.mime}; codecs="${codec.codec}"`;
				if (oracles.canPlay(type) !== '' || (await oracles.decodingSupported(side.mime, codec.codec))) {
					tokens.push(`${container.token}/${codec.token}`);
				}
			}
		}
	}
	return tokens;
}

let probed: Promise<string> | null = null;

/*
 * The capability list to send on a plan request, probed once per page load
 * and cached: the matrix is fixed and a browser's decoders do not change
 * under it mid-session. An empty list is a valid answer — the client
 * claims nothing and the server keeps its conservative rules — so an
 * environment without a document (a test runner) probes nothing.
 */
export function reportCapabilities(): Promise<string> {
	probed ??= probe();
	return probed;
}

async function probe(): Promise<string> {
	if (typeof document === 'undefined' || typeof navigator === 'undefined') return '';
	const media = document.createElement('video');
	const tokens = await tokensFor({
		canPlay: (type) => media.canPlayType(type),
		decodingSupported: decodingSupported
	});
	return tokens.join(',');
}

async function decodingSupported(mime: string, codec: string | null): Promise<boolean> {
	const capabilities = navigator.mediaCapabilities as MediaCapabilities | undefined;
	if (!capabilities || typeof capabilities.decodingInfo !== 'function') return false;
	const contentType = codec === null ? mime : `${mime}; codecs="${codec}"`;
	const config: MediaDecodingConfiguration = mime.startsWith('video/')
		? {
				type: 'file',
				video: { contentType, width: 1920, height: 1080, bitrate: 8_000_000, framerate: 24 }
			}
		: { type: 'file', audio: { contentType, channels: '2', bitrate: 128_000, samplerate: 48_000 } };
	try {
		const info = await capabilities.decodingInfo(config);
		return info.supported;
	} catch {
		// A configuration the implementation refuses to parse is a "no",
		// not a crash: the whole list is an optimisation, never a
		// prerequisite for playback.
		return false;
	}
}

File: internal/contracts/capabilities.go
package contracts

import (
	"sort"
	"strings"
)

// ClientCapabilities is a client's claim about what it can decode. It is
// deliberately not a verdict about one file: the plan is requested
// before the client knows the file's streams, so the client reports what
// it has verified it can decode and the planner applies that claim to
// the probed streams (D-058).
//
// Tokens are lowercase and come in two shapes:
//
//	"mkv"      the container itself opens
//	"mkv/h264" the codec family decodes inside that container
//
// The pairing is not cosmetic. A browser can decode HEVC through a
// platform decoder inside MP4 and still fail on HEVC inside Matroska —
// measured here as a black screen with no error event, an advancing
// clock and zero decoded frames. A flat codec list would license exactly
// that, so a codec token is only ever claimed for a container the client
// also opened.
//
// A claim never buys more than the plan decision: the player verifies
// the first decoded frame and falls back to a transcode when none
// arrives (D-058), and only tokens this server knows are acted on.
type ClientCapabilities struct {
	Tokens []string `json:"tokens"`
}

// capabilityContainers maps the file extensions this server serves
// directly to the container token a client reports for them. An
// extension absent here is never direct-played, whatever a client
// claims: the gateway would have no honest Content-Type for it.
var capabilityContainers = map[string]string{
	"mp4": "mp4", "m4v": "mp4", "mov": "mp4", "m4a": "mp4",
	"mkv":  "mkv",
	"webm": "webm",
	"ogg":  "ogg", "ogv": "ogg", "oga": "ogg", "opus": "ogg",
	"mp3":  "mp3",
	"flac": "flac",
}

// capabilityCodecs maps the ffprobe codec names that reach this server
// to the family token a client reports for them. Codecs absent here
// (pcm_*, mpeg4 part 2, ...) have no token at all, so no client can
// claim them and they always remux or transcode.
var capabilityCodecs = map[string]string{
	"h264": "h264", "avc1": "h264",
	"hevc": "hevc", "h265": "hevc",
	"vp8": "vp8", "vp9": "vp9", "av1": "av1",
	"aac": "aac", "mp3": "mp3", "opus": "opus", "vorbis": "vorbis",
	"flac": "flac", "ac3": "ac3", "eac3": "eac3", "dts": "dts",
}

// DirectContainer returns the container token for a file extension
// (".mkv" and "mkv" are both accepted). The second result is false when
// this server has no directly playable container for that extension.
func DirectContainer(ext string) (string, bool) {
	container, ok := capabilityContainers[strings.ToLower(strings.TrimPrefix(ext, "."))]
	return container, ok
}

// CapabilityCodec returns the codec family token for an ffprobe codec
// name. The second result is false for codecs no browser decodes.
func CapabilityCodec(codec string) (string, bool) {
	family, ok := capabilityCodecs[strings.ToLower(codec)]
	return family, ok
}

// KnownCapabilityToken reports whether a token is inside this server's
// vocabulary. Everything else is dropped on the way in, so a client
// cannot widen the set the planner reasons about.
func KnownCapabilityToken(tok string) bool {
	container, codec, paired := strings.Cut(tok, "/")
	if !knownContainerToken(container) {
		return false
	}
	if !paired {
		return true
	}
	for _, family := range capabilityCodecs {
		if family == codec {
			return true
		}
	}
	return false
}

func knownContainerToken(tok string) bool {
	for _, container := range capabilityContainers {
		if container == tok {
			return true
		}
	}
	return false
}

// ParseCapabilities normalizes a raw client capability list — one query
// value, or several comma-separated ones — into the set the planner
// acts on: lowercase, deduplicated, sorted, and stripped of tokens this
// server does not know. Unknown tokens are dropped rather than rejected:
// the list is a claim, and a client from a future version must not have
// its request fail for naming something this server cannot use.
//
// A nil result means unknown (nothing was reported) and the planner
// keeps its conservative rules. A non-nil result with no tokens means
// the client claims it can decode nothing, which is a decision.
func ParseCapabilities(raw []string) *ClientCapabilities {
	if len(raw) == 0 {
		return nil
	}
	seen := make(map[string]bool, len(raw))
	tokens := make([]string, 0, len(raw))
	for _, value := range raw {
		for _, tok := range strings.Split(value, ",") {
			tok = strings.ToLower(strings.TrimSpace(tok))
			if tok == "" || seen[tok] || !KnownCapabilityToken(tok) {
				continue
			}
			seen[tok] = true
			tokens = append(tokens, tok)
		}
	}
	sort.Strings(tokens)
	return &ClientCapabilities{Tokens: tokens}
}

// Has reports whether the client claimed a token. A nil capability set
// claims nothing; callers that must tell "unknown" from "empty" check
// for nil before asking.
func (c *ClientCapabilities) Has(token string) bool {
	if c == nil {
		return false
	}
	for _, tok := range c.Tokens {
		if tok == token {
			return true
		}
	}
	return false
}
